src/browser/history.js

```diff
@@ -0,0 +1,321 @@
+
+  Tag.remove = (tag, item) =>
+    tag.update('items', exclude(item[PouchDB.id]));
+
+  Tag.tag = (item, tagName) => item.update('tags', include(tagName));
+
+  Tag.untag = (item, tagName) => item.update('tags', exclude(tagName));
+
+  const Top = Record({
+    [PouchDB.id]: ID('top/pages'),
+    [PouchDB.revision]: Maybe(Revision),
+    pages: List(Page)
+  });
+  Top.storeID = Page.storeID;
+  Top.sample = (page, limit) => top => top.update('pages', pages => {
+    const index = pages.findIndex(x => x[PouchDB.id] === page[PouchDB.id])
+    return pages.set(index < 0 ? pages.size : index, page)
+                .sortBy(Page.frequency)
+                .take(limit);
+  });
+
+  const clear = async(function*({stores}) {
+    yield stores.pages.destroy();
+    yield stores.quotes.destroy();
+    yield stores.tags.destroy();
+  });
+
+  let taskID = 0;
+
+  const scheduleEdit = async(function*(history, record, transform) {
+    try {
+      // wait for last scehduled edit to complete.
+      yield history.editQueue[record[PouchDB.id]];
+    } catch (error) {
+      // If last operation faild log an error, but proceed to a next
+      // operation.
+      console.error(error);
+    } finally {
+      const store = history.stores[record.constructor.storeID];
+      return edit(record, transform, store);
+    }
+  })
+
+  // History
+  class History {
+    static defaults() {
+      return {
+        pagesStoreName: "pages",
+        quotesStoreName: "quotes",
+        tagsStoreName: "tags",
+        topPageLimit: 6,
+        trackTopPages: false
+      }
+    }
+    constructor(options={}) {
+      this.onPageChange = this.onPageChange.bind(this);
+      this.onTopPagesChange = this.onTopPagesChange.bind(this);
+
+      this.options = Object.assign(History.defaults(), options);
+      const {pagesStore, quotesStore, tagsStore,
+             pagesStoreName, quotesStoreName, tagsStoreName} = this.options;
+
+      this.stores = {
+        pages: pagesStore || new PouchDB(pagesStoreName),
+        quotes: quotesStore || new PouchDB(quotesStoreName),
+        tags: tagsStore || new PouchDB(tagsStoreName)
+      }
+
+      this.editQueue = Object.create(null);
+
+      if (options.trackTopPages) {
+        this.setupChangeFeeds();
+        this.setupListeners();
+      }
+    }
+    setupChangeFeeds() {
+      const {pages} = this.stores
+
+      this.topPageChangeFeed = pages.changes({
+        since: "now",
+        live: true,
+        include_docs: true,
+        doc_ids: ["top/pages"]
+      });
+
+      this.pagesChangeFeed = pages.changes({
+        since: "now",
+        live: true,
+        filter: ({[PouchDB.id]: id}) => id.startsWith("page/"),
+        include_docs: true
+      });
+    }
+    setupListeners() {
+      this.topPageChangeFeed.on("change", this.onTopPagesChange);
+      this.pagesChangeFeed.on("change", this.onPageChange);
+    }
+
+    // Edits per record are queued, to avoid data loss
+    // due to concurrent edits.
```

Comment:
> When, if ever, do we have concurrent edits on the client side?
> 
> Also, doesn't supplying a _rev cause Pouch to automatically handle these things for us? Seems like this is exactly what Couch/Pouch are designed to do. Please feel free to tell me I am wrong if I'm missing something here.

Here is an example:

1. Load `facebook.com`
2. Create `facebook.com` page record.
3. Get title change event edit that records `title`.
4. Get change icon event edit records `icon`.

Last three steps make edits to the same record and 4th edit may start before 3rd is complete in which case both 3rd and 4th operation will refer to the `_rev` from 2nd edit and there for run into conflict and get rejected.

In practice though we load many tabs at the same time and some of them maybe for the same page and even if not each edit triggers update on `top/sites`. So different page record do get updated concurrently with `top/sites` and there for without queueing edits you